warmrestart/warmRestartHelper.cpp

```diff
@@ -0,0 +1,362 @@
+ * one received from the application, and will proceed to push it down to AppDB.
+ */
+void WarmStartHelper::reconcile(void)
+{
+    SWSS_LOG_NOTICE("Warm-Restart: Initiating reconciliation process for %s "
+                    "application.", m_appName.c_str());
+
+    assert(getState() == WarmStart::RESTORED);
+
+    for (auto &restoredElem : m_restorationVector)
+    {
+        std::string restoredKey  = kfvKey(restoredElem);
+        auto restoredFV          = kfvFieldsValues(restoredElem);
+
+        auto iter = m_refreshMap.find(restoredKey);
+
+        /*
+         * If the restored element is not found in the refreshMap, we must
+         * push a delete operation for this entry.
+         */
+        if (iter == m_refreshMap.end())
+        {
+            SWSS_LOG_NOTICE("Warm-Restart reconciliation: deleting stale entry %s",
+                            printKFV(restoredKey, restoredFV).c_str());
+
+            m_syncTable->del(restoredKey);
+            continue;
+        }
+
+        /*
+         * If an explicit delete request is sent by the application, process it
+         * right away.
+         */
+        else if (kfvOp(iter->second) == DEL_COMMAND)
+        {
+            SWSS_LOG_NOTICE("Warm-Restart reconciliation: deleting entry %s",
+                            printKFV(restoredKey, restoredFV).c_str());
+
+            m_syncTable->del(restoredKey);
+        }
+
+        /*
+         * If a matching entry is found in refreshMap, proceed to compare it
+         * with its restored counterpart.
+         */
+        else
+        {
+            auto refreshedKey = kfvKey(iter->second);
+            auto refreshedFV  = kfvFieldsValues(iter->second);
+
+            if (compareAllFV(restoredFV, refreshedFV))
+            {
+                SWSS_LOG_NOTICE("Warm-Restart reconciliation: updating entry %s",
+                                printKFV(refreshedKey, refreshedFV).c_str());
+
+                m_syncTable->set(refreshedKey, refreshedFV);
+            }
+            else
+            {
+                SWSS_LOG_INFO("Warm-Restart reconciliation: no changes needed for "
+                              "existing entry %s",
+                              printKFV(refreshedKey, refreshedFV).c_str());
+            }
+        }
+
+        /* Deleting the just-processed restored entry from the refreshMap */
+        m_refreshMap.erase(restoredKey);
+    }
+
+    /*
+     * Iterate through all the entries left in the refreshMap, which correspond
+     * to brand-new entries to be pushed down to AppDB.
+     */
+    for (auto &kfv : m_refreshMap)
+    {
+        auto refreshedKey = kfvKey(kfv.second);
+        auto refreshedFV  = kfvFieldsValues(kfv.second);
+
+        SWSS_LOG_NOTICE("Warm-Restart reconciliation: introducing new entry %s",
+                        printKFV(refreshedKey, refreshedFV).c_str());
+
+        m_syncTable->set(refreshedKey, refreshedFV);
```

Comment:
Seems to me that if you received a delete for a prefix not in the db, you will be creating a stale prefix.

Reply:
@nikos-github This route 'delete', which could only arrive to fpmsyncd if sent right after an 'add', would be discarded right away at orchagent level as it's missing nh attributes, so it wouldn't ever make it to asicdb, but i agree that we should stop it at fpmsyncd level. Will make a small correction in this logic. Thanks for pointing this out.

Reply:
If the request is a delete operation,  `kfvOp(iter->second) == DEL_COMMAND`,  it won't reach here.

Reply:
@jipanyang , see that Niko's point is for the scenario in which the deleted-route is not present in AppDB, so we won't hit this code that you're highlighting.

Reply:
@rodnymolina @jipanyang Please read carefully the case I have described. If the prefix is not in the db (and hence not restored), a route delete for which the route add was missed since the connection was severed, will result in a stale prefix being added and never removed. The logic and code are wrong at this point.

Reply:
I see.   Actually it is not an issue with current sonic.  consumerStateTable will automatically transform key without field/value into a del operation, no stale prefix will be added. 

But it is better to do the check in WarmStartHelper library instead of relying on consumerStateTable.

Reply:
done